src/main/kotlin/org/jetbrains/research/testspark/actions/TestSparkAction.kt

```diff
@@ -352,15 +351,53 @@ class TestSparkAction : AnAction() {
                         )
                     }
                 } else {
-                    // TODO: show the warning panel
+                    // TODO shouw the warning panel
                 }
             }
             visibilityController.isVisible = false
             dispose()
         }
 
-        private fun startEvoSuiteGeneration() = startUnitTestGenerationTool(tool = EvoSuite())
-        private fun startLLMGeneration() = startUnitTestGenerationTool(tool = Llm())
+        private fun startLLMGeneration() {
+            if (!testGenerationController.isGeneratorRunning(project)) {
+                val testSamplesCode = llmSampleSelectorFactory.getTestSamplesCode()
+
+                if (psiHelper != null) {
+                    if (codeTypeButtons[0].isSelected) {
+                        Llm().generateTestsForClass(
+                            project,
+                            psiHelper,
+                            caretOffset,
+                            fileUrl,
+                            testSamplesCode,
+                            testGenerationController,
+                        )
+                    } else if (codeTypeButtons[1].isSelected) {
+                        Llm().generateTestsForMethod(
+                            project,
+                            psiHelper,
+                            caretOffset,
+                            fileUrl,
+                            testSamplesCode,
+                            testGenerationController,
+                        )
+                    } else if (codeTypeButtons[2].isSelected) {
+                        Llm().generateTestsForLine(
+                            project,
+                            psiHelper,
+                            caretOffset,
+                            fileUrl,
+                            testSamplesCode,
+                            testGenerationController,
+                        )
+                    }
+                } else {
+                    // TODO shouw the warning panel
+                }
+            }
+            visibilityController.isVisible = false
+            dispose()
+        }
```

Comment:
It looks like the code duplication removed in #244 ([PR](https://github.com/JetBrains-Research/TestSpark/pull/244/files)) got introduced back. Why aren't the changes from the PR applied?

Reply:
Also, I think it does not really relates to Psi classes